File: typescript/src/core/modules/auth/functions/verify-token.ts
import jwt from "jsonwebtoken";
import jose from "node-jose";

// Import constants
import { Configs } from "../../../../utils/configs/index.js";

// Import errors
import { AppError, ClientError, isStandardError } from "../../../error";

import { initializeInternalContext } from "../../../context/internal-context/index.js";

// Import helpers
import { getAuthorizationToken } from "../helpers/get-authorization-token.js";
import { getPublicKeys } from "../helpers/get-public-keys.js";

// Import types
import type { JwtPayload } from "jsonwebtoken";
import type { TRuntimeContext } from "../../../context/runtime-context/type.js";
import type { Pipeline } from "../../../context/pipeline/index.js";

const { JWK } = jose;

const appClientId = Configs.CognitoAppClientId;

/**
 * Xác thực token xem có hợp lệ hay không?
 *
 * @param ctx - runtime context.
 *
 * @returns
 */
export async function verifyToken(ctx: TRuntimeContext) {
  try {
    // Setup context before go further
    const internalCtx = initializeInternalContext();
    (internalCtx.params as any).headers = await ctx.getHeaders();
    internalCtx.options!.canCatchError = true;

    const token = getAuthorizationToken(internalCtx)!;
    const decodedHeader = jwt.decode(token, { complete: true });

    if (!decodedHeader || !decodedHeader.header.kid) {
      throw new Error("Invalid token header");
    }

    const keys = await getPublicKeys();

    if (isStandardError(keys)) {
      return keys;
    }

    const key = keys.find((key: any) => key.kid === decodedHeader.header.kid);

    if (!key) {
      const err = new AppError("Token is invalid");
      err.addErrorDetail({
        source: "verifyToken",
        desc: "Public key not found",
      });
      err.asHTTPError("InternalServerError");
      return err;
    }

    const jwkKey = await JWK.asKey({
      kty: key.kty,
      n: key.n,
      e: key.e,
    });

    const publicKey = jwkKey.toPEM();
    const claims = jwt.verify(token, publicKey, {
      algorithms: ["RS256"],
    }) as JwtPayload;

    // Post check claims
    if (Date.now() / 1000 > claims.exp!) {
      const err = new ClientError("Token is invalid");
      err.addErrorDetail({
        source: "verifyToken",
        desc: "Token is expired",
      });
      err.asHTTPError("Unauthorized");
      return err;
    }

    if (claims.client_id !== appClientId) {
      const err = new ClientError("Token is invalid");
      err.addErrorDetail({
        source: "verifyToken",
        desc: "Token was not issued for this audience",
      });
      err.asHTTPError("Unauthorized");
      return err;
    }

    return claims;
  } catch (error: any) {
    if (isStandardError(error)) return error;

    const err = new AppError("Verify token failed");
    err.asHTTPError("InternalServerError");
    err.addErrorDetail({ source: "verifyToken", desc: error.message });
    return err;
  }
}

/**
 * Tạo ra một step executor có gán pipeline trong đó để xác thực token.
 *
 * @param pipeline - pipeline
 *
 * @returns
 */
export function createVerifyTokenStepExecutor(pipeline: Pipeline<any>) {
  return async function (ctx: TRuntimeContext) {
    const result = await verifyToken(ctx);

    if (isStandardError(result)) {
      // Stop pipeline
      pipeline.stop(ctx);

      return ctx.sendError(result);
    }

    // Save claims as temp data
    ctx.addTempData("claims", result);

    return result;
  };
}

File: python/src/core/modules/auth/functions/verify_token.py
# Import built-in libraries
import time

# Import external libraries
import jwt
from jwcrypto import jwk

# Import from core
from core.error import is_standard_error, AppError
from core.context.internal_context import initialize_internal_context

# Import from utils
from core.error import is_standard_error
from utils.configs import Configs

# Import helpers
from core.modules.auth.helpers import (
    get_authorization_token,
    get_public_keys,
    get_info_from_claims,
)


def verify_token(ctx):
    """Xác thực token và trả về kết quả xác thực và claims (nếu thành công).

    Args:
        token (str): token

    Raises:
        ValueError: ném ra lỗi nếu như header của JWT Token là không hợp lệ
        ValueError: ném ra lỗi nếu như không tìm thấy Public Key
        ValueError: ném ra lỗi nếu như token được tạo ra không phải là từ
        Cognito Provider nguồn

    Returns:
        dict: kết quả xác thực
    """
    try:
        # Setup context before go further
        in_ctx = initialize_internal_context()
        in_ctx.params["headers"] = ctx.get_headers()
        in_ctx.options["can_catch_error"] = True

        token = get_authorization_token(ctx)

        keys = get_public_keys()

        if is_standard_error(keys):
            return keys

        decoded_header = jwt.get_unverified_header(token)
        kid = decoded_header.get("kid")

        if not kid:
            err = AppError("Token is invalid")
            err.add_error_detail(
                {"source": "verify_token", "desc": "Kid of token not found"}
            )
            err.as_http_error("Unauthorized")
            return err

        key = next((k for k in keys if k["kid"] == kid), None)
        if not key:
            err = AppError("Token is invalid")
            err.add_error_detail(
                {"source": "verify_token", "desc": "Public key not found"}
            )
            err.as_http_error("Unauthorized")
            return err

        jwk_key = jwk.JWK(**key)
        public_key_pem = jwk_key.export_to_pem()

        claims = jwt.decode(
            token,
            public_key_pem,
            algorithms=["RS256"],
            options={"verify_exp": True},
        )

        # Post check claims
        if int(time.time()) > claims.exp:
            err = AppError("Token is invalid")
            err.add_error_detail({"source": "verify_token", "desc": "Token is expired"})
            err.as_http_error("Unauthorized")
            return err

        if claims.get("client_id") != Configs.Cognito_App_Client_Id:
            err = AppError("Token is invalid")
            err.add_error_detail(
                {
                    "source": "verify_token",
                    "desc": "Token was not issued for this audience",
                }
            )
            err.as_http_error("Unauthorized")
            return err

        return claims

    except Exception as e:
        if is_standard_error(e):
            return e

        err = AppError("Verify token failed")
        err.add_error_detail({"source": "verify_token", "desc": str(e)})
        err.as_http_error("InternalServerError")
        return err


def create_verify_token_step_executor(pipeline):
    """Tạo ra một step executor có gán pipeline trong đó để xác thực token.

    Args:
        class: pipeline

    Returns:
        Callable: executor của step.
    """

    def executor(ctx):
        result = verify_token(ctx)

        if is_standard_error(result):
            # Stop pipeline
            pipeline.stop()

            return ctx.sendError(result)

        ctx.add_temp_data("claims", result)

        return result

    return executor
